File: src/functions/bot-join/botjoin.go
package botjoin

import (
	"bytes"
	"net/http"
	"regexp"
	"text/template"

	"github.com/roleypoly/roleypoly/src/common"
	"github.com/roleypoly/roleypoly/src/common/faas"
)

var (
	validGuildID         = regexp.MustCompile(`^[0-9]+$`)
	redirectPathTemplate = template.Must(
		template.New("redirect").Parse(
			`https://discord.com/api/oauth2/authorize?client_id={{.ClientID}}&scope=bot&permissions={{.Permissions}}{{if .GuildID}}&guild_id={{.GuildID}}&disable_guild_select=true{{end}}`,
		),
	)
	clientID = common.Getenv("BOT_CLIENT_ID").String()
)

type redirectPathData struct {
	ClientID    string
	Permissions int
	GuildID     string
}

func BotJoin(rw http.ResponseWriter, r *http.Request) {
	guildID := r.URL.Query().Get("guild")
	if !validGuildID.MatchString(guildID) {
		guildID = ""
	}

	pathData := redirectPathData{
		ClientID:    clientID,
		Permissions: 268435456, // MANAGE_ROLES
		GuildID:     guildID,
	}

	pathBuffer := bytes.Buffer{}
	redirectPathTemplate.Execute(&pathBuffer, pathData)

	faas.Bounce(rw, pathBuffer.String())
}

File: src/backend-worker/handlers/bot-join.ts
import { Bounce } from '../utils/bounce';

const validGuildID = /^[0-9]+$/;

type URLParams = {
    clientID: string;
    permissions: number;
    guildID?: string;
};

const buildURL = (params: URLParams) => {
    let url = `https://discord.com/api/oauth2/authorize?client_id=${params.clientID}&scope=bot&permissions=${params.permissions}`;

    if (params.guildID) {
        url += `&guild_id=${params.guildID}&disable_guild_select=true`;
    }

    return url;
};

export const BotJoin = (request: Request): Response => {
    let guildID = new URL(request.url).searchParams.get('guild') || '';

    if (guildID && !validGuildID.test(guildID)) {
        guildID = '';
    }

    return Bounce(
        buildURL({
            clientID: BOT_CLIENT_ID,
            permissions: 268435456,
            guildID,
        })
    );
};
